pkg/ui/react-app/src/thanos/pages/blocks/helpers.ts

```diff
@@ -103,3 +103,28 @@ export const download = (blob: Block): string => {
 
   return url;
 };
+
+export const getOverlappingBlocks = (blockPools: { [source: string]: BlocksPool }): Set<string> => {
+  let result: Set<string> = new Set();
+
+  Object.keys(blockPools).forEach((pk) => {
+    let blocks = Object.values(blockPools[pk]);
+    for (let i = 0; i < blocks.length; i++) {
+      if (blocks[i].length <= 1) {
+        continue;
+      }
+      for (let j = 0; j < blocks[i].length - 1; j++) {
+        for (let k = 0; k < blocks[i][j].length; k++) {
+          for (let l = 0; l < blocks[i][j + 1].length; l++) {
+            if (isOverlapping(blocks[i][j][k], blocks[i][j + 1][l])) {
```

Comment:
the computational complexity is not thaaaat bad. The fact that everything is so nested is because the data structure is so complex, but it's not soooo far off from O(N). It's definitely not O(N^2), as that would mean comparing every block in every row in every block pool to every other block. We are (only) comparing every block within one row of a single block pool to all of the other blocks in the same row of the same block pool, which is much less. Maybe we can still simplify this.